### PR TITLE
Support backup replica read

### LINKED ISSUE #14060: Support replica read for backup
## Development Task
Support replica read for backup

### PR DESCRIPTION
**Pull request by v01dstar**
Signed-off-by: Yang Zhang <yang.zhang@pingcap.com>

### What is changed and how it works?

Issue Number: Close #14060

What's Changed:

```commit-message
Support replica read for backup
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Unit test

Side effects

n/a

### Release note 

```release-note
Support replica read for backup
```

### CODE REVIEW

File: components/backup/src/endpoint.rs [hunk 1]
```diff
@@ -769,18 +783,26 @@ impl<R: RegionInfoProvider> Progress<R> {
                             break;
                         }
                     }
-                    if info.role == StateRole::Leader {
+                    let peer = find_peer(region, store_id).unwrap().to_owned();
+                    // Raft peer role has to match the replica read flag.
+                    if (replica_read == BackupReplicaRead::Leader && info.role == StateRole::Leader)
+                        || (replica_read == BackupReplicaRead::Follower
+                            && info.role != StateRole::Leader)
```
Comment (sticnarf) [inline components/backup/src/endpoint.rs:790]: Need we check it so strictly? In other words, is it really necessary to distinguish between `BackupReplicaRead::Follower` and `BackupReplicaRead::LeaderAndFollower`?
If a peer just becomes leader but it receives a follower read request, I think we needn't reject it in this case.

Comment (v01dstar) [reply components/backup/src/endpoint.rs:790]: In most cases, `BackupReplicaRead::Follower`, `BackupReplicaRead::LeaderAndFollower` means nearly the same to the user. But, in some cases, for example, a user prefers all the leaders are in certain zones, while they also want the backup to run in other zones, then these 2 options make difference.

Comment (sticnarf) [reply components/backup/src/endpoint.rs:790]: > But, in some cases, for example, a user prefers all the leaders are in certain zones, while they also want the backup to run in other zones, then these 2 options make difference.

To my understanding, the user is perfectly clear about the zones where the TiKV stores are. However, leaders are less stable than stores themselves. Although we try to put all leaders in a certain zone, temporary network isolation or server crash may cause the leader to move to other zones.

So, in this case, the user can easily guarantee they don't send requests to a remote zone, but they cannot 100% guarantee the roles of the peers in the zone.

Comment (v01dstar) [reply components/backup/src/endpoint.rs]: I removed `FOLLOWER` only option

**Review by sticnarf — COMMENTED** [review]
LGTM

File: components/backup/src/endpoint.rs [hunk 2]
```diff
@@ -769,18 +783,24 @@ impl<R: RegionInfoProvider> Progress<R> {
                             break;
                         }
                     }
-                    if info.role == StateRole::Leader {
+                    let peer = find_peer(region, store_id).unwrap().to_owned();
```
Comment (3pointer) [inline components/backup/src/endpoint.rs:786]: Will both leader and follower backup the same data in different stores when (replica_read == BackupReplicaRead::LeaderAndFollower)?

Comment (v01dstar) [reply components/backup/src/endpoint.rs:786]: Indeed, we need the client to make sure that only one replica receives the backup request. Considering TiDB backup currently "push down" backup requsts to all stores, checking on the server side cannot guarantee exclusivity. Will change this.

**Review by sticnarf — APPROVED** [review]

File: components/backup/src/endpoint.rs [hunk 3]
```diff
@@ -769,18 +783,20 @@ impl<R: RegionInfoProvider> Progress<R> {
...
-                        let leader = find_peer(region, store_id).unwrap().to_owned();
                         let backup_range = BackupRange {
                             start_key: skey,
                             end_key: ekey,
                             region: region.clone(),
-                            leader,
+                            peer,
                             codec,
                             cf: cf_name,
+                            uses_replica_read: info.role != StateRole::Leader,
```
Comment (3pointer) [inline components/backup/src/endpoint.rs:799]:
```suggestion
                            uses_replica_read: replica_read,
```

File: components/backup/src/endpoint.rs [hunk 4]
```diff
@@ -304,35 +307,45 @@ impl BackupRange {
...
+                            IsolationLevel::Si,
+                        )
+                    },
+                )
+                .map_err(MvccError::from)
+                .map_err(TxnError::from)?;
+        }
 
         let start_snapshot = Instant::now();
         let snapshot = match engine.snapshot(snap_ctx) {
```
Comment (hicqu) [inline components/backup/src/endpoint.rs:351]: Can it success on region follower?

File: components/backup/src/endpoint.rs (hunk 3 above)
Comment (3pointer) [reply components/backup/src/endpoint.rs:799]: It seems no problem

**Review by hicqu — APPROVED** [review]

**Review by hicqu — APPROVED** [review]

**Pull request merged**